New Exodus IGA Tests PR

roystgnr opened the pull request:
This is the same changeset as in #19415 but hopefully will get put through newer CI recipes

roystgnr committed:
Add initial IGA test

This first mesh file has no sidesets, so we'll do a transient solve, to
get a decent run with only Neumann BCs.

Refs #18768

roystgnr committed:
Add 3d IGA test

roystgnr committed:
Better test requirement messages

roystgnr committed:
clear_spline_nodes param in FileMesh, FM Generator

This lets us choose to load an IsoGeometric Analysis mesh without spline
nodes and spline node constraints, whether for debugging purposes or to
experiment with solver behavior or just because libMesh doesn't
parallelize those constraints efficiently enough yet.

Refs #18768

roystgnr committed:
Add IGA test with spline nodes cleared

This works for me but let's make sure it doesn't break later.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/996009/) on 38b12f9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20646/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/996017/) on 38b12f9 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7e4878</th>
      <th colspan=4>#20646 38b12f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.55%</td>
      <td>82.55%</td>
      <td>+0.00%</td>
      <td>87.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73710</td>
      <td>73719</td>
      <td>+9</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15580</td>
      <td>15579</td>
      <td>-1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20646/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20646/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20646/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 87.50% is less than the suggested 90.0%



This comment will be updated on new commits.

fdkong commented on test/tests/meshgenerators/file_mesh_generator/tests:
```diff
@@ -22,6 +22,42 @@
+  []
+
+  [./exodus_pseudoiga_test]
+    type = 'RunApp'
```
Should we do exodiff? I would expect we could un IGA for simple problems such as diffusion and got the correct results. Are we not there yet?

roystgnr replied:
We don't have proper output yet.  The CoreForm people have been working on improving libMesh VTK output for IGA meshes, but I don't know if it's there yet or if we can diff it if so, and right now if we output Exodus our RATIONAL_BERNSTEIN elements will be treated like LAGRANGE and the geometry distorted.  I don't want to put in a test employing that.

roystgnr replied:
It's an interesting question, whether we should disable imperfect output types of IGA meshes entirely in the meantime.  Exodus IGA output is still useful (sometimes as a "close enough in the eyeball norm" way to visualize things, sometimes as a "look at the control points for their own sake" tool) though.

aeslaughter replied:
I think doing VTK output with VTKDiff would work, as far as I know it is all working as expected.

roystgnr replied:
CI agrees, when we have VTK 9; I'll just need to disable the tests for VTK 8 or lower.

permcody commented:
Looks reasonable to me. However, on our call last week you mentioned some potential errors that should occur if we use the wrong mesh or perhaps just bad combination of options (what happens when you change "SECOND" to "FIRST" in any of these tests? Do you get a reasonable error message? 

Also, I don't imagine we'd care if somebody cleared spline nodes on our normal first order meshes, do we just silently do nothing in that case? Just thinking about all the bad ways that users (or ignorant developers like myself) can use this incorrectly and get poor error messages or worse, bad runs (is that even possible)?

Feel free to use the `RunException` type tests if you'd like to test for and maintain specific error messages.

cticenhour reviewed (changes requested):
Leaving a reminder about adding to the news so we can highlight this work.

roystgnr commented:
> what happens when you change "SECOND" to "FIRST" in any of these tests? Do you get a reasonable error message?

Not in the slightest!  `Assertion 'comp < this->n_comp(s,var)' failed.` in dbg mode (with 0 < 0, so I can actually tell at a glance what that implies, but I'd be surprised+saddened if anyone else can), and `Matrix is missing diagonal 11` in opt mode (which is beyond my powers of comprehension).

The trouble here is I'm not actually sure how to improve on that.

Compare when we try to run too-high-order elements on first-order meshes: change QUAD9 to QUAD4 and run test_hier_2_2d, for example, and dbg mode gives us a decent error, but opt mode dies at a confusing `Quadrature rule 5 not supported!`

There's a bit of tension in both cases between "we shouldn't let error-checking slow down the simulation" and "we shouldn't let the user get themselves in trouble" when the error-checking is necessarily `O(N_elem*N_evaluations)`.

> Also, I don't imagine we'd care if somebody cleared spline nodes on our normal first order meshes, do we just silently do nothing in that case?

If there's no spline nodes (on a first-order or on a higher-order Lagrange-mapped mesh), yeah, clearing them is a no-op.

> Just thinking about all the bad ways that users (or ignorant developers like myself) can use this incorrectly and get poor error messages

This is a longstanding mess.  I'm just not sure how to fix it without adding a billion extra branch instructions.  "Add them anyway and then benchmark it", I guess?  Maybe if I only added more error checking to the DoF numbering pass, not to every single FE evaluation, the cost would be negligible.

> or worse, bad runs (is that even possible)?

This hurts, but: probably, technically?  If we've made it far enough to hit a preconditioner error message, we're probably in a state where either we've already committed Undefined Behavior or we're practically begging PETSc to do so.  I'd expect 99.9999% of the time we get a segfault at worst, but a dangling pointer dereference corrupting memory isn't out of the question.

I'll look into "Add them anyway and then benchmark it".  That'll probably have to be at the libMesh level, though, not something I can slip into this PR.

roystgnr committed:
Require VTK 9 for XMLDiff of IGA VTK files vs gold

With VTK 8 our IGA output files are incomplete, and getting an output
diff tested with some build configurations is probably better than just
getting a converged/didn't-converge result tested with all
configurations.

roystgnr committed:
Support vtk version checks in TestHarness

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001230/) on b2e3881 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001230/) on b2e3881 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001230/) on b2e3881 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

roystgnr committed:
Disallow renumbering in IGA tests

roystgnr committed:
parallel_type = replicated for IGA tests

I'm very confused here.  We're disallowing renumbering, and we're on
only one processor, yet we get XML diffs when running on
DistributedMesh?  All the libMesh IGA tests work just fine on
distributed meshes...

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001342/) on 237e5c3 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001342/) on 237e5c3 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001342/) on 237e5c3 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

roystgnr commented:
> The following file(s) are changed:
>
> conda/libmesh/meta.yaml

Why does moosebuild lie?

loganharbour commented:
> > The following file(s) are changed:
> > conda/libmesh/meta.yaml
> 
> Why does moosebuild lie?

The logic for finding the merge base (we take devel and pull your PR) when your PR contains something from next is flawed. I just haven’t gotten around to fixing it

fdkong reviewed (approved):
Looks nice to me

cticenhour reviewed (approved):
News was added in #20652

roystgnr commented on test/tests/meshgenerators/file_mesh_generator/2d_diffusion_iga.i:
```diff
@@ -0,0 +1,52 @@
+
+[Variables]
+  [u]
+    order = SECOND  # Must match mesh order
```
See https://github.com/idaholab/moose/pull/20646#issuecomment-1084979056 - the answer is "no", which is awful, but also "that's already a problem with non-IGA runs" and "there's no good way to fix that at the Moose level".  I'm hoping you'll agree that puts it out of the scope of this PR?

fdkong replied:
Is it possible to get the mesh order after the entire mesh is already loaded? If so, we should be able to compare the FEM order with the mesh order after the mesh being loaded.

In the current state, I do feel a little sad on this. I am pretty sure dump Fande will spend several hours on debugging if he specifies a wrong FEM order for a given mesh before realizing that what he should do

roystgnr replied:
There's no such thing as "the mesh order" - users are mixing different elem orders in the same mesh in some codes.  Plus, it'd be a full memory sweep's extra expense, and worse still it'd give us false positives; "load a mesh then run it through `all_second_order()`" is a very useful workflow.  libMesh gives users loads of flexibility to change things right up until the last minute, which unfortunately makes it hard to be sure they aren't going to fix an apparent inconsistency right before the last minute.

I think the right fix then is "test at the very last minute" - I'll work up something that inspects elements as we're assigning DoF indices, so we can throw errors with more informative messages than the low-level errors we otherwise hit.  But again, that's a method in libMesh, not Moose.  And even if it was a fix to go in at the Moose level, would this be the right place to put it?  Surely it'd be a bit much to block a new feature addition unless it simultaneously fixes a 10-year-old UI inadequacy?

fdkong replied:
I am fine if we want to do this in a follow-up PR
